src/arena.c

```diff
@@ -717,129 +759,175 @@ arena_decay_time_set(tsdn_t *tsdn, arena_t *arena, ssize_t decay_time) {
 	 * arbitrary change during initial arena configuration.
 	 */
-	arena_decay_reinit(&arena->decay, &arena->extents_cached, decay_time);
-	arena_maybe_purge(tsdn, arena, &arena->decay, &arena->extents_cached);
-	malloc_mutex_unlock(tsdn, &arena->decay.mtx);
+	arena_decay_reinit(decay, extents, decay_time);
+	arena_maybe_decay(tsdn, arena, decay, extents);
+	malloc_mutex_unlock(tsdn, &decay->mtx);
 
 	return false;
 }
 
+bool
+arena_dirty_decay_time_set(tsdn_t *tsdn, arena_t *arena, ssize_t decay_time) {
+	return arena_decay_time_set(tsdn, arena, &arena->decay_dirty,
+	    &arena->extents_dirty, decay_time);
+}
+
+bool
+arena_muzzy_decay_time_set(tsdn_t *tsdn, arena_t *arena, ssize_t decay_time) {
+	return arena_decay_time_set(tsdn, arena, &arena->decay_muzzy,
+	    &arena->extents_muzzy, decay_time);
+}
+
 static size_t
-arena_stash_dirty(tsdn_t *tsdn, arena_t *arena, extent_hooks_t **r_extent_hooks,
-    extents_t *extents, size_t ndirty_limit, extent_list_t *purge_extents) {
+arena_stash_decayed(tsdn_t *tsdn, arena_t *arena,
+    extent_hooks_t **r_extent_hooks, extents_t *extents, size_t npages_limit,
+    extent_list_t *decay_extents) {
 	witness_assert_depth_to_rank(tsdn, WITNESS_RANK_CORE, 0);
 
-	/* Stash extents according to ndirty_limit. */
+	/* Stash extents according to npages_limit. */
 	size_t nstashed = 0;
 	extent_t *extent;
 	while ((extent = extents_evict(tsdn, arena, r_extent_hooks, extents,
-	    ndirty_limit)) != NULL) {
-		extent_list_append(purge_extents, extent);
+	    npages_limit)) != NULL) {
+		extent_list_append(decay_extents, extent);
 		nstashed += extent_size_get(extent) >> LG_PAGE;
 	}
 	return nstashed;
 }
 
 static size_t
-arena_purge_stashed(tsdn_t *tsdn, arena_t *arena,
-    extent_hooks_t **r_extent_hooks, extent_list_t *purge_extents) {
-	UNUSED size_t nmadvise;
+arena_decay_stashed(tsdn_t *tsdn, arena_t *arena,
+    extent_hooks_t **r_extent_hooks, arena_decay_t *decay, extents_t *extents,
+    bool all, extent_list_t *decay_extents) {
+	UNUSED size_t nmadvise, nunmapped;
 	size_t npurged;
 
 	if (config_stats) {
 		nmadvise = 0;
+		nunmapped = 0;
 	}
 	npurged = 0;
 
-	for (extent_t *extent = extent_list_first(purge_extents); extent !=
-	    NULL; extent = extent_list_first(purge_extents)) {
+	ssize_t muzzy_decay_time = arena_muzzy_decay_time_get(arena);
+	for (extent_t *extent = extent_list_first(decay_extents); extent !=
+	    NULL; extent = extent_list_first(decay_extents)) {
 		if (config_stats) {
 			nmadvise++;
 		}
-		npurged += extent_size_get(extent) >> LG_PAGE;
-		extent_list_remove(purge_extents, extent);
-		extent_dalloc_wrapper(tsdn, arena, r_extent_hooks, extent);
+		size_t npages = extent_size_get(extent) >> LG_PAGE;
+		npurged += npages;
+		extent_list_remove(decay_extents, extent);
+		switch (extents_state_get(extents)) {
+		case extent_state_active:
+			not_reached();
+		case extent_state_dirty:
+			if (!all && muzzy_decay_time != 0 &&
```

Comment:
This checks for ```muzzy_decay_time != 0``` so that we fall through to immediate deallocation or forced purge (whatever ```extent_dalloc_wrapper()``` does) if the muzzy decay time is 0, i.e. "decay immediately".